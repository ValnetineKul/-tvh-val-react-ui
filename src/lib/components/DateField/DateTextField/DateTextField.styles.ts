import { makeStyles } from '../../../../themes/core';
import { surface } from '../../TextField/TextField.constants';

const useStyles = makeStyles({ name: 'DateTextField' })((theme) => ({
  adornmentInputRoot: {
    width: '100%',
    '& input': {
      height: 24,
      ...theme.font['screen/body500/regular/default'],
      padding: `calc(${theme.layoutSpacing['spacing/400']} + ${theme.layoutSpacing['spacing/200']}) ${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/200']} ${theme.layoutSpacing['spacing/000']} !important`,
      color: theme.color[`text/onSurface${surface}/default`],

      '&:disabled': {
        WebkitTextFillColor: theme.color[`text/onSurface${surface}/disabled`],
      },
      '&::placeholder': {
        color: `${theme.color[`text/onSurface${surface}/placeholder`]} !important`,
        opacity: 1,
      },
    },
    '& fieldset': {
      border: 'none',
    },
  },

  input: {
    '& input': {
      padding: '0px !important',
    },
  },

  visuallyHidden: {
    clip: 'rect(0 0 0 0)',
    clipPath: 'inset(50%)',
    height: 1,
    overflow: 'hidden',
    position: 'absolute',
    whiteSpace: 'nowrap',
    width: 1,
  },
}));

export default useStyles;
