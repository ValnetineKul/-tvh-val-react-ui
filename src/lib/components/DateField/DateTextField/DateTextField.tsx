import React, { useState } from 'react';
import type { FC } from 'react';
import { default as MUIDesktopDatePicker } from '@mui/lab/DesktopDatePicker';
import { default as MUITextField } from '@mui/material/TextField';
import type { InputBaseComponentProps, InputBaseProps } from '@mui/material/InputBase';
import { default as MuiInputBase } from '@mui/material/InputBase';
import Icon from '../../Icon';
import { Calendar } from '../../Icon/icons/functional';
import type { TextFieldProps } from '../../TextField';
import TextField from '../../TextField';
import type { DataAttributes } from '../../../types/common';
import useStyles from './DateTextField.styles';

export interface DateTextFieldProps {
  format?: string;
  isDateRange?: boolean;
  adornmentInputValue?: Date | null;
  startInputId?: string;
  endInputId?: string;
  setIsInputFocused: (isFocused: boolean) => void;
  onAdornmentInputValueChange: (dateValue: Date | null) => void;
  inputPropsDateRangeTo?: InputBaseComponentProps & DataAttributes;
  inputPropsDateRangeFrom?: InputBaseComponentProps & DataAttributes;
}

const InputComponent = React.forwardRef<HTMLInputElement>(
  (props: { isDateRange?: boolean; label?: string } & InputBaseProps & DataAttributes, ref) => {
    const { classes, cx } = useStyles();

    const { isDateRange, label, color, id, className, ...remainingProps } = props;

    return (
      <>
        <label className={classes.visuallyHidden} htmlFor={id}>
          {isDateRange ? 'End date' : label}
        </label>
        <MuiInputBase
          {...remainingProps}
          id={id}
          ref={ref}
          className={cx(classes.adornmentInputRoot, classes.input, className)}
        />
      </>
    );
  }
);

const DateTextField: FC<TextFieldProps & DateTextFieldProps> = ({
  format,
  adornmentInputValue,
  setIsInputFocused,
  onAdornmentInputValueChange,
  isDateRange,
  inputPropsDateRangeTo,
  inputPropsDateRangeFrom,
  label,
  focused,
  disabled,
  readOnly,
  inputProps,
  startInputId,
  endInputId,
  ...props
}) => {
  const { classes } = useStyles();
  const [isAdornmentInputFocused, setIsAdornmentInputFocused] = useState(false);

  const handleOnAdornmentInputFocus = (isFocused: boolean) => {
    setIsAdornmentInputFocused(isFocused);
    setIsInputFocused(isFocused);
  };

  return (
    <TextField
      {...props}
      type="text"
      id={isDateRange ? endInputId : label}
      placeholder={format?.toLowerCase()}
      label={label}
      InputLabelProps={{ component: 'span', htmlFor: undefined }}
      focused={focused || isAdornmentInputFocused}
      disabled={disabled}
      readOnly={readOnly}
      startIcon={<Icon icon={Calendar} />}
      inputComponent={InputComponent}
      inputProps={{ ...inputProps, ...inputPropsDateRangeTo, readOnly, isDateRange, label }}
      onFocus={() => setIsInputFocused(true)}
      onBlur={() => setIsInputFocused(false)}
      startAdornment={
        isDateRange ? (
          <MUIDesktopDatePicker
            disableOpenPicker
            inputFormat={format}
            open={false}
            disabled={disabled}
            value={adornmentInputValue}
            onChange={onAdornmentInputValueChange}
            renderInput={(textFieldProps) => {
              const { type, ...remainingInputProps } = textFieldProps.inputProps;
              return (
                <MUITextField
                  {...textFieldProps}
                  id={startInputId}
                  autoComplete="off"
                  placeholder={format?.toLowerCase()}
                  label="Start date"
                  InputLabelProps={{
                    shrink: true,
                    className: classes.visuallyHidden,
                    htmlFor: startInputId,
                    id: undefined,
                  }}
                  onFocus={() => handleOnAdornmentInputFocus(true)}
                  onBlur={() => handleOnAdornmentInputFocus(false)}
                  className={classes.adornmentInputRoot}
                  inputProps={{ ...remainingInputProps, ...inputPropsDateRangeFrom, readOnly }}
                />
              );
            }}
          />
        ) : null
      }
    />
  );
};

export default DateTextField;
