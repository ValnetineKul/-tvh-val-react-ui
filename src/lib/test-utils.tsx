import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
