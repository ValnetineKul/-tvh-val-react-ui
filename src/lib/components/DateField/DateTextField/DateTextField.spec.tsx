import React from 'react';
import { render } from '../../../test-utils';
import DateTextField from './DateTextField';

describe('DateTextField', () => {
  it('Should render correctly', () => {
    const { container } = render(
      <DateTextField
        format="dd/MM/yyyy"
        setIsInputFocused={() => {}}
        adornmentInputValue={new Date('03/10/2022')}
        onAdornmentInputValueChange={() => {}}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
