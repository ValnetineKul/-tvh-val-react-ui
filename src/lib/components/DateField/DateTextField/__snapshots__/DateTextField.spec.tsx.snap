// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTextField Should render correctly 1`] = `
<div
  class="MuiFormControl-root tss-10-TextField-withStartIcon-ref tss-10-TextField-withStartAdornment-ref tss-ubelmr-TextField-root-TextField-withStartIcon-TextField-withStartAdornment mui-1nrlq1o-MuiFormControl-root"
>
  <div
    class="tss-bdw984-Surface-root"
  >
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart tss-10-TextField-inputRoot-ref tss-11lw9i6-TextField-inputRoot-TextField-adornedStart mui-ha1n7c-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium tss-10-TextField-adornment-ref tss-1mcfx9-InputAdornment-root-TextField-adornment-TextField-startIcon mui-ittuaa-MuiInputAdornment-root"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
        <div
          class="tss-w1sjks-Icon-root-Icon-sizeMd"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
            color="inherit"
            focusable="false"
            viewBox="0 0 24 24"
          >
            calendar.svg
          </svg>
        </div>
      </div>
      <label
        class="tss-gx6y4z-DateTextField-visuallyHidden"
      />
      <div
        aria-invalid="false"
        class="MuiInputBase-root MuiInputBase-colorPrimary MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart tss-18k3wr1-DateTextField-adornmentInputRoot-DateTextField-input-TextField-input mui-15nd16p-MuiInputBase-root-MuiInputBase-input-MuiOutlinedInput-input"
      >
        <input
          class="MuiInputBase-input mui-yz9k0d-MuiInputBase-input"
          placeholder="dd/mm/yyyy"
          type="text"
          value=""
        />
      </div>
      <fieldset
        aria-hidden="true"
        class="tss-tyiogj-TextField-notchedOutline tss-10-TextField-notchedOutline-ref MuiOutlinedInput-notchedOutline mui-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="mui-nnbavb"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="tss-19z6szl-TextField-helpers"
  />
</div>
`;
